ee/tools/mdm/certs.go

```diff
@@ -0,0 +1,257 @@
+package main
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+
+	"github.com/micromdm/micromdm/pkg/crypto/mdmcertutil"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	vendorCertEnvName = "VENDOR_CERT_PEM"
+	vendorKeyEnvName  = "VENDOR_KEY_PEM"
+	vendorPassEnvName = "VENDOR_KEY_PASSPHRASE"
+	csrEnvName        = "CSR_BASE64"
+
+	rsaPrivateKeyPEMBlockType = "RSA PRIVATE KEY"
+	certificatePEMBlockType   = "CERTIFICATE"
+	csrPEMBlockType           = "CERTIFICATE REQUEST"
+
+	wwdrIntermediaryURL = "https://www.apple.com/certificateauthority/AppleWWDRCAG3.cer"
+	appleRootCAURL      = "https://www.apple.com/appleca/AppleIncRootCertificate.cer"
+)
+
+// emailAddressOID defined by https://oidref.com/1.2.840.113549.1.9.1
+var emailAddressOID = []int{1, 2, 840, 113549, 1, 9, 1}
+
+// See
+// https://developer.apple.com/documentation/devicemanagement/implementing_device_management/setting_up_push_notifications_for_your_mdm_customers
+// for the expected CSR format.
+
+func main() {
+	// Load vendor keys and certs
+	vendorCertPEM := os.Getenv(vendorCertEnvName)
+	if vendorCertPEM == "" {
+		log.Fatalf("vendor cert must be set in %s", vendorCertEnvName)
+	}
+	vendorKeyPEM := os.Getenv(vendorKeyEnvName)
+	if vendorKeyPEM == "" {
+		log.Fatalf("vendor key must be set in %s", vendorKeyEnvName)
+	}
+	vendorKeyPassphrase := os.Getenv(vendorPassEnvName)
+	if vendorKeyPassphrase == "" {
+		log.Fatalf("vendor key passphrase must be set in %s", vendorPassEnvName)
+	}
+	vendorCert, err := decodeVendorCert([]byte(vendorCertPEM))
+	if err != nil {
+		log.Fatalf("failed to parse vendor cert: %s", err.Error())
+	}
+	vendorKey, err := loadKey([]byte(vendorKeyPEM), []byte(vendorKeyPassphrase))
+	if err != nil {
+		log.Fatalf("failed to load vendor private key: %s", err.Error())
+	}
+
+	// Decode CSR input
+	// We accept the CSR via environment variable to mitigate against command injection attacks in
+	// the fleetdm.com website code that will call this with untrusted user input.
+	csrBase64 := os.Getenv(csrEnvName)
+	if csrBase64 == "" {
+		log.Fatalf("CSR must be set in %s", csrEnvName)
+	}
+	csr, err := base64.StdEncoding.DecodeString(string(csrBase64))
+	if err != nil {
+		log.Fatalf("base64 decode csr: %s", err.Error())
+	}
+	certReq, err := decodeCSR(csr)
+	if err != nil {
+		log.Fatalf("decode pem: %s", err.Error())
+	}
+
+	// Get email from CSR
+	email, err := getEmail(certReq)
```

Comment:
In addition to the email, we are also including the organization in the csr in https://github.com/fleetdm/fleet/pull/8812. Should we extract this and include it in the response?

Reply:
just connecting threads for posterity here, seems like this is also being discussed here: https://github.com/fleetdm/fleet/pull/8812#discussion_r1035394426

Reply:
Yes, it sounds like collecting the org is a requirement, so I will add that.